### DISCUSSION TITLE
Ubuntu 22: Wayland aborts without a connected monitor. No hotplug either + Xorg weird crash

### DISCUSSION THREAD

**Discussion by marc-hb** [thesofproject/sof-test#964] [General]
Key discovery:

  => Wayland does NOT run without a monitor attached! Half of our test devices don't have a monitor...

```
Sep 29 22:26:15 jf-adlp-rvp-nocodec-5 org.gnome.Shell.desktop[4517]: Failed to setup: No GPUs with outputs found
```

Wayland is the default in Ubuntu 22 (not available in Ubuntu 20).

Wayland (or rather: the specific Wayland implementation shipped in Ubuntu 22) does not support monitor hotplug either! Wayland stops when you unplug the monitor and does NOT restart when you plug it back. A manual `sudo systemctl restart gdm3` does restart wayland after a monitor hotplug.

An aborted Wayland never shows up in `systemctl --user --failed` or `systemctl --anything-I-tried`; you have to run `ps faux` to see whether Wayland failed to start or not.

When no display is attached, some devices automatically fallback on Xorg at boot time while other devices give up and don't start gdm3 at all. I don't know why yet; I don't know what makes this difference. For the Xorg fallback you must have `apt install xserver-xorg-core` of course but that's the default, it's part of `apt install ubuntu-desktop`. Again: having `xserver-xorg-core` is not always enough to trigger a fallback to Xorg and I don't know why.

Whether it's by choice or because of a fallback, Xorg processes run ~fine~ even without a display, just like in the older Ubuntu 20.04 days. (EDIT: not actually that simple and good, see comment below.) Xorg still supports hotplug too, no change either.  Sometimes Xorg without a display shows up and complains in `systemctl --user --failed` (which is what was caught in our validation in the first place) and other times not. I don't know why (EDIT: mystery solved, see comment below) Example: https://sof-ci.01.org/sofpr/PR6342/build1659/devicetest/?model=TGLU_UP_HDA_ZEPHYR&testcase=verify-kernel-boot-log
```
$ systemctl --user --wait is-system-running
degraded
$ systemctl --user --failed
  UNIT                        LOAD   ACTIVE SUB    DESCRIPTION
● org.gnome.Shell@x11.service loaded failed failed GNOME Shell on X11
```

```
Sep 29 22:28:41 up2 gnome-shell[1147]: JS ERROR: TypeError: monitor is null
                                       _updateWorkAreaBox@resource:///org/gnome/shell/ui/overviewControls.js:58:26
                                       _init@resource:///org/gnome/shell/ui/overviewControls.js:53:14
                                       ControlsManagerLayout@resource:///org/gnome/shell/ui/overviewControls.js:31:1
                                       _init@resource:///org/gnome/shell/ui/overviewControls.js:388:31
                                       ControlsManager@resource:///org/gnome/shell/ui/overviewControls.js:307:1
                                       _init@resource:///org/gnome/shell/ui/overview.js:77:26
                                       OverviewActor@resource:///org/gnome/shell/ui/overview.js:65:1
                                       init@resource:///org/gnome/shell/ui/overview.js:212:26
                                       _initializeUI@resource:///org/gnome/shell/ui/main.js:246:14
                                       start@resource:///org/gnome/shell/ui/main.js:175:5
                                       @resource:///org/gnome/shell/ui/init.js:6:17
Sep 29 22:28:41 up2 gnome-shell[1147]: Execution of main.js threw exception: Module resource:///org/gnome/shell/ui/init.js threw an exception
Sep 29 22:28:41 up2 gnome-shell[1147]: Attempting to run a JS callback during garbage collection. This is most likely caused by destroying a Clutter actor or GTK widget with ::destroy s>
Sep 29 22:28:41 up2 gnome-shell[1147]: == Stack trace for context 0x55c1157ac180 ==
Sep 29 22:28:41 up2 gnome-shell[1147]: The offending callback was set_actor(), a vfunc.
Sep 29 22:28:41 up2 gnome-shell[1147]: Attempting to call back into JSAPI during the sweeping phase of GC. This is most likely caused by not destroying a Clutter actor or Gtk+ widget wi>
Sep 29 22:28:41 up2 gnome-shell[1147]: The offending signal was destroy on StEntry 0x55c116fbd860.
Sep 29 22:28:41 up2 gnome-shell[1147]: == Stack trace for context 0x55c1157ac180 ==
Sep 29 22:28:41 up2 gnome-shell[1147]: == Stack trace for context 0x55c1157ac180 ==
Sep 29 22:28:41 up2 gnome-shell[1147]: == Stack trace for context 0x55c1157ac180 ==
Sep 29 22:28:41 up2 gnome-shell[1147]: Attempting to call back into JSAPI during the sweeping phase of GC. This is most likely caused by not destroying a Clutter actor or Gtk+ widget wi>
Sep 29 22:28:41 up2 gnome-shell[1147]: The offending signal was destroy on StEntry 0x55c116fbd860.
Sep 29 22:28:41 up2 gnome-shell[1147]: == Stack trace for context 0x55c1157ac180 ==
Sep 29 22:28:41 up2 gnome-shell[1147]: == Stack trace for context 0x55c1157ac180 ==
Sep 29 22:28:41 up2 gnome-shell[1147]: == Stack trace for context 0x55c1157ac180 ==

```
3 times, then:
```
Sep 29 22:28:41 up2 systemd[714]: org.gnome.Shell@x11.service: Failed with result 'protocol'.
Sep 29 22:28:41 up2 systemd[714]: Failed to start GNOME Shell on X11.
Sep 29 22:28:41 up2 systemd[714]: org.gnome.Shell@x11.service: Consumed 2.268s CPU time.
Sep 29 22:28:41 up2 systemd[714]: org.gnome.Shell@x11.service: Scheduled restart job, restart counter is at 3.
Sep 29 22:28:41 up2 systemd[714]: Stopped GNOME Shell on X11.
Sep 29 22:28:41 up2 systemd[714]: org.gnome.Shell@x11.service: Consumed 2.268s CPU time.
Sep 29 22:28:41 up2 systemd[714]: org.gnome.Shell@x11.service: Start request repeated too quickly.
Sep 29 22:28:41 up2 systemd[714]: org.gnome.Shell@x11.service: Failed with result 'protocol'.
Sep 29 22:28:41 up2 systemd[714]: Failed to start GNOME Shell on X11.
```

--------

So how do you switch between Wayland and Xorg? (for instance because you want monitor hotplug back?)

First, you must understand that this is a per-user preference. You can choose when you enter your password at login time, see this screenshot: https://askubuntu.com/questions/968827/no-option-to-switch-between-x11-and-wayland-on-ubuntu-17-10

Your choice is persisted in the file `/var/lib/AccountsService/users/<username>` like this:
```
[User]
Session=ubuntu-xorg
```

Not having this file is OK and it defaults to Wayland[*].

But wait: what runs started before you even enter your password? Wayland or Xorg? There's no user yet so no preference yet. By default it's Wayland but you can override it in `/etc/gdm3/custom.conf` (for instance: because you want hotplug back?)

But wait again: once someone has switched from Wayland to Xorg, there is no way back! Wayland can start Xorg but Xorg never switches back to Wayland. The choice is removed for future users after you logged off, everyone must use Xorg until the next reboot! This can lead to very funny situations like monitor hotplug never works before anyone logged in but then hotplug mysteriously works after some (Xorg) user logged in at least once. But hotplug still does not work if only Wayland users have logged in before.

For even more fun, note we use "autologin"  in /etc/gdm3/custom.conf on our all DUTs. I'll spare you the extra fun this can cause :-)

[*] when there is no `/var/lib/AccountsService/users/<username>` file, some devices automatically recreate a template/empty file at some point, other devices not. I don't what makes this difference. As expected, this template does not seem to have any effect, still defaults to Wayland.

Just for reference:
https://docs.fedoraproject.org/en-US/quick-docs/configuring-xorg-as-default-gnome-session/

**Comment by fredoh9** [comment]
One silly but simple question is, can we just un-install Wayland and set Xorg as default? Then no need to deal with Wayland fun stuff?

**Reply by marc-hb** [reply]
It is a possibility. Uninstallation is not even required, `Session=ubuntu-xorg` is enough (see above).

However:
- we would test an unusual system configuration and...
- would it make any audio-related difference? Maybe, maybe not - needs more research.
- We would be affected by other, non-audio problems that no one else has, not just `failed GNOME Shell on X11`. For instance on one of the systems I tested org.gnome.SettingsDaemon.Power.desktop crashed every time I logged in with Xorg. No crash with Wayland. No idea why.
- We would merely postpone the inevitable? Unless Ubuntu+GNOME+Wayland add some support for headless systems in the future.
- EDIT: Xorg does not really seem to support headless systems either, see more recent comment below.

**Comment by marc-hb** [comment]
Actually, **Xorg does not really support headless systems either**: it partly crashes when there's no display. That's what `systemctl --user --failed` has been trying to tell us the whole time. 

> Sometimes Xorg without a display shows up and complains in systemctl --user --failed (which is what was caught in our validation in the first place) and other times not. I don't know why.

.. and I just found why: with `AutomaticLoginEnable` in `/etc/gdm3/custom.conf` + no monitor then Xorg partly crashes and that crash gets reported in `systemctl --user --failed`; but ONLY when someone logs in. Before anyone logs in then `systemctl` is still clean.

If there's a monitor then Xorg does not crash.

One of the things that confused me is that `ps faux` still shows some Xorg process when there is no monitor + automaticlogin but are they all there and in what state? Not a good one according to `systemctl`. Does hotplug really work? More testing needed.

```
Sep 29 22:28:41 up2 gnome-shell[1147]: JS ERROR: TypeError: monitor is null

Sep 29 22:28:41 up2 systemd[714]: org.gnome.Shell@x11.service: Failed with result 'protocol'.
Sep 29 22:28:41 up2 systemd[714]: Failed to start GNOME Shell on X11.


Sep 29 22:28:41 up2 systemd[714]: org.gnome.Shell@x11.service: Start request repeated too quickly.
Sep 29 22:28:41 up2 systemd[714]: org.gnome.Shell@x11.service: Failed with result 'protocol'.
Sep 29 22:28:41 up2 systemd[714]: Failed to start GNOME Shell on X11.

systemctl --user --failed
  UNIT                        LOAD   ACTIVE SUB    DESCRIPTION
● org.gnome.Shell@x11.service loaded failed failed GNOME Shell on X11
```

So another possibility is to turn off the GUI entirely (BOTH wayland and Xorg) with `systemctl set-default multi-user`?

**Comment by marc-hb** [comment]
To check whether the kernel driver has detected a display or not:
```
$ sudo grep -i connected /sys/kernel/debug/dri/0/i915_display_info
 
[CONNECTOR:236:eDP-1]: status: connected
[CONNECTOR:245:DP-1]: status: disconnected
[CONNECTOR:256:HDMI-A-1]: status: disconnected
[CONNECTOR:262:DP-2]: status: disconnected
[CONNECTOR:271:DP-3]: status: disconnected
[CONNECTOR:280:DP-4]: status: disconnected
[CONNECTOR:289:DP-5]: status: disconnected
```

EDIT, another one that does NOT require root.

```
$ DISPLAY=:0 xrandr --listmonitors
Monitors: 1
 0: +*HDMI-1 1920/509x1080/286+0+0  HDMI-1
```
```
$ DISPLAY=:0 xrandr --listmonitors
Monitors: 1
 0: +XWAYLAND0 2560/600x1440/340+0+0  XWAYLAND0
```

It also shows when `gdm3` is not started:

```shell
$ sudo systemctl isolate multi-user
DISPLAY=:0 xrandr --listmonitors
Can't open display :0

$ sudo systemctl isolate default

```

**Comment by marc-hb** [comment]
I borrowed and tried a couple "HDMI dummy plugs" and they make all these problems just go away.

**Reply by marc-hb** [reply]
Thanks @kv2019i for sharing that HDMI audio capture devices also solve this problem. Kills two birds with one stone - this is an audio project here after all :-)

**Comment by marc-hb** [comment]
> `DISPLAY=:0 xrandr --listmonitors`

This does not work any more with Ubuntu 22 over ssh:

EDIT: it works again as of October 2023?!

```
$ DISPLAY=:0 xrandr --listmonitors
Can't open display :0

$ DISPLAY=:1024 xrandr --listmonitors
Authorization required, but no authorization protocol specified
Can't open display :1024

```

This leaves:
```
sudo grep -i connected /sys/kernel/debug/dri/0/i915_display_info
```

**Comment by marc-hb** [comment]
gdm fails in an even more spectacular way when there's no GPU: #998

**Comment by marc-hb** [comment]
Shared privately by Jason:
> You can use x11 without a monitor if you use a “dummy” video driver (no special HDMI dummy plug needed).   I run all my headless servers this way with xvnc as the vnc-server.  I can remote-connect straight into a full x11 desktop environment on these machines.

**Reply by marc-hb** [reply]
This is especially important because some issues occur only when the display is ON:
- https://github.com/thesofproject/linux/issues/4861

See useful troubleshooting commands there (turning display on and off, etc.)
